### PR TITLE
New opd forms

### PR DESCRIPTION
**Pull request by hamza-vd**
New OPD forms implementation

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -31,4 +31,6 @@ android.useAndroidX=true
 android.enableD8=true
 android.enableD8.desugaring=true
 android.enableJetifier=true
-android.jetifier.blacklist=shadows
\ No newline at end of file
+android.jetifier.blacklist=shadows
+sonatypeUsername=opensrp
+sonatypePassword=b6Ap_sWUpr+wUqa6@br_sUDrebRa=r7W
```
Comment (qiarie) [inline gradle.properties:36]: Please move this to `local.properties` and clean these credentials from git history

File: opensrp-opd/src/main/java/org/smartregister/opd/repository/VisitDetailsRepository.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        if (visitDetail.getUpdatedAt() != null)
+            values.put(UPDATED_AT, visitDetail.getUpdatedAt().getTime());
+        return values;
+    }
+
+    public void addVisitDetails(VisitDetail visitDetail) {
+        addVisitDetails(visitDetail, getWritableDatabase());
+    }
+
+    public void addVisitDetails(VisitDetail visitDetail, SQLiteDatabase database) {
```
Comment (bennsimon) [inline opensrp-opd/src/main/java/org/smartregister/opd/repository/VisitDetailsRepository.java:72]: Why pass database as an argument?

File: opensrp-opd/src/main/java/org/smartregister/opd/utils/VisitUtils.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+            CheckInEventProcessException {
+        String tableName = OpdUtils.metadata().getTableName();
+
+        String lastInteractedWithDate = String.valueOf(new Date().getTime());
+
+        ContentValues contentValues = new ContentValues();
+        contentValues.put("last_interacted_with", lastInteractedWithDate);
+
+        int recordsUpdated = OpdLibrary.getInstance().getRepository().getWritableDatabase()
+                .update(tableName, contentValues, "base_entity_id = ?", new String[]{event.getBaseEntityId()});
```
Comment (bennsimon) [inline opensrp-opd/src/main/java/org/smartregister/opd/utils/VisitUtils.java:126]: Move `currentObject.getEventDate()` and `base_entity_id` to constants.

File: opensrp-opd/src/main/java/org/smartregister/opd/utils/VisitUtils.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+            return false;
+        }
+        return true;
+    }
+
+        public static String getFormattedDate(SimpleDateFormat source_sdf, SimpleDateFormat dest_sdf, String value) {
+        try {
+            Date date = source_sdf.parse(value);
+            return dest_sdf.format(date);
+        } catch (Exception e) {
```
Comment (bennsimon) [inline opensrp-opd/src/main/java/org/smartregister/opd/utils/VisitUtils.java:198]:
```suggestion
        } catch (ParseException e) {
```

Or use a less generic exception.

File: opensrp-opd/src/main/java/org/smartregister/opd/utils/VisitUtils.java [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        dateFormat.setLenient(false);
+        try {
+            dateFormat.parse(inDate.trim());
+        } catch (ParseException pe) {
+            return false;
+        }
+        return true;
+    }
+
+        public static String getFormattedDate(SimpleDateFormat source_sdf, SimpleDateFormat dest_sdf, String value) {
```
Comment (bennsimon) [inline opensrp-opd/src/main/java/org/smartregister/opd/utils/VisitUtils.java]: Kindly follow convention in variable naming `source_sdf` and `dest_sdf`

File: opensrp-opd/src/main/java/org/smartregister/opd/utils/VisitUtils.java [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        try {
+            Date date = source_sdf.parse(value);
+            return dest_sdf.format(date);
+        } catch (Exception e) {
+            try {
+                // fallback for long datetypes
+                Date date = new Date(Long.parseLong(value));
+                return dest_sdf.format(date);
+            } catch (NumberFormatException | NullPointerException nfe) {
+                Timber.e(e);
```
Comment (bennsimon) [inline opensrp-opd/src/main/java/org/smartregister/opd/utils/VisitUtils.java:204]:
```suggestion
                Timber.e(nfe);
```

File: opensrp-opd/src/test/java/org/smartregister/opd/fragment/NewOpdProfileOverviewFragmentTest.java [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        HashMap<String, List<ProfileAction.ProfileActionVisit>> mapVisits = new HashMap<>();
+        ArrayList<ProfileAction.ProfileActionVisit> visits = new ArrayList<>();
+        ProfileAction.ProfileActionVisit visit = new ProfileAction.ProfileActionVisit();
+        visit.setVisitID("abc123");
+        visits.add(visit);
+        mapVisits.put("OPD_treatment", visits);
+
+        fragment.loadGlobals(mapVisits);
+        HashMap<String, String> formGlobalValues = ReflectionHelpers.getField(fragment, "formGlobalValues");
+        Assert.assertEquals(formGlobalValues.size(), globalKeys.size());
```
Comment (bennsimon) [inline opensrp-opd/src/test/java/org/smartregister/opd/fragment/NewOpdProfileOverviewFragmentTest.java:94]: swap arguments

**Review by bennsimon — CHANGES_REQUESTED** [review]

File: opensrp-opd/src/main/java/org/smartregister/opd/fragment/NewOpdProfileOverviewFragment.java [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+    }
+
+    public void getGlobalKeysFromConfig() {
+        try {
+            Iterable<Object> opdGlobals = OpdLibrary.getInstance().readYaml(OpdConstants.FileUtils.OPD_GLOBALS, new Yaml());
+            for (Object ruleObject : opdGlobals) {
+                Map<String, Object> map = ((Map<String, Object>) ruleObject);
+                globalKeys.addAll((List<String>) map.get(JsonFormConstants.FIELDS));
+            }
+        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (hamza-vd) [reply opensrp-opd/src/main/java/org/smartregister/opd/fragment/NewOpdProfileOverviewFragment.java]: Changed to `IOException`

File: opensrp-opd/src/main/java/org/smartregister/opd/utils/VisitUtils.java (hunk 3 above)
Comment (hamza-vd) [reply opensrp-opd/src/main/java/org/smartregister/opd/utils/VisitUtils.java]: moved `base_entity_id` and `last_interacted_with` to constants

File: opensrp-opd/src/main/java/org/smartregister/opd/utils/VisitUtils.java (hunk 5 above)
Comment (hamza-vd) [reply opensrp-opd/src/main/java/org/smartregister/opd/utils/VisitUtils.java]: Changed to `sourceDateFormat` and `destDateFormat` respectively

File: opensrp-opd/src/main/java/org/smartregister/opd/fragment/NewOpdProfileVisitsFragment.java [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+                            formGlobalValues.put(globalKey, "");
+                        }
+                    }
+                    if (savedValues.containsKey(OpdConstants.REPEATING_GROUP_MAP)) {
+                        String testResults = OpdJsonFormUtils.getLabResultsStringFromMap(savedValues);
+                        formGlobalValues.put("diagnostic_test_lab_results", testResults);
+                    }
+
+                }
+                formGlobalValuesMap.put(key, formGlobalValues);
```
[the remark this answers is not in the document]
Comment (paulinembabu) [reply opensrp-opd/src/main/java/org/smartregister/opd/fragment/NewOpdProfileVisitsFragment.java:118]: These sections of code are different

File: opensrp-opd/src/main/java/org/smartregister/opd/repository/VisitRepository.java [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+                    + LOCATION_ID + " VARCHAR NULL, "
+                    + CHILD_LOCATION_ID + " VARCHAR NULL, "
+                    + VISIT_GROUP + " VARCHAR NULL, "
+                    + BASE_ENTITY_ID + " VARCHAR NULL, "
+                    + VISIT_DATE + " VARCHAR NULL, "
+                    + FORM_SUBMISSION_ID + " VARCHAR NULL, "
+                    + UPDATED_AT + " DATETIME NULL, "
+                    + CREATED_AT + " DATETIME NULL, "
+                    + DELETED_AT + " DATETIME NULL)";
+    private static final String BASE_ENTITY_ID_INDEX = "CREATE INDEX " + VISIT_TABLE + "_" + BASE_ENTITY_ID + "_index ON " + VISIT_TABLE
```
[the remark this answers is not in the document]
Comment (paulinembabu) [reply opensrp-opd/src/main/java/org/smartregister/opd/repository/VisitRepository.java:48]: Visit Type is the Event Type name, Created at is being used to show the time the event was created, Visit group is the date the event was done. For easy grouping

**Review by paulinembabu — DISMISSED** [review]

File: opensrp-opd/src/main/java/org/smartregister/opd/repository/VisitRepository.java [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+            Timber.e(e);
+        } finally {
+            if (cursor != null) {
+                cursor.close();
+            }
+        }
+        return (visits.size() > 0) ? visits.get(0) : null;
+    }
+
+    public String getLastInteractedWithAndVisitNotDone(String baseEntityID, String dateColumn) {
```
[the remark this answers is not in the document]
Comment (paulinembabu) [reply opensrp-opd/src/main/java/org/smartregister/opd/repository/VisitRepository.java:265]: Not sure where this comment applies

**Review by paulinembabu — APPROVED** [review]

**Pull request merged**